File: vendors/waitrose.py
from playwright.async_api import Page

from .base import (
    Cart, CartItem, Order, OrderDetail, OrderItem, Product, ShoppingVendor,
)

_BASE = 'https://www.waitrose.com'
_GRAPHQL = f'{_BASE}/api/graphql-prod/graph/live'

# Reused across fetch calls and the 401-retry path
_FETCH_JS = """async ([method, url, body, token]) => {
    const opts = { method, headers: { 'Content-Type': 'application/json' } };
    if (token) opts.headers['Authorization'] = token;
    if (body !== null) opts.body = JSON.stringify(body);
    const r = await fetch(url, opts);
    const text = await r.text();
    let data;
    try { data = JSON.parse(text); } catch { data = text; }
    return { status: r.status, body: data };
}"""


class WaitroseVendor(ShoppingVendor):
    def __init__(self, page: Page):
        self._page = page
        self._customer_id = ''
        self._order_id = ''
        self._token = ''

    async def _init_context(self):
        ctx = await self._page.evaluate("""() => {
            // Hook window.fetch to capture tokens from the SPA's own outgoing requests.
            if (!window.__shopmeHooked) {
                window.__shopmeHooked = true;
                const orig = window.fetch;
                window.fetch = function(input, init) {
                    init = init || {};
                    const h = init.headers || {};
                    const auth = (typeof h.get === 'function')
                        ? h.get('authorization')
                        : (h['authorization'] || h['Authorization'] || '');
                    if (auth && auth.startsWith('Bearer ')) window.__shopmeToken__ = auth;
                    return orig.call(this, input, init);
                };
            }
            const ctx = {
                customerId: localStorage.getItem('wtr_customer_id') || '',
                orderId:    localStorage.getItem('wtr_order_id')    || '',
                token:      window.__shopmeToken__                  || '',
            };
            // Scan SSR script elements — token and orderId live there on page load
            // even after the SPA overwrites window.__PRELOADED_STATE__ with true.
            for (const script of document.querySelectorAll('script')) {
                const text = script.textContent;
                if (!ctx.token) {
                    const m = text.match(/"accessToken":"(Bearer [^"]+)"/);
                    if (m) ctx.token = m[1];
                }
                if (!ctx.orderId) {
                    const m = text.match(/"customerOrderId":"([^"]+)"/);
                    if (m) ctx.orderId = m[1];
                }
                if (!ctx.customerId) {
                    const m = text.match(/"customerId":"([^"]+)"/);
                    if (m) ctx.customerId = m[1];
                }
                if (ctx.token && ctx.orderId && ctx.customerId) break;
            }
            return ctx;
        }""")
        self._customer_id = ctx.get('customerId') or ''
        self._order_id    = ctx.get('orderId')    or ''
        self._token       = ctx.get('token')      or ''

        # If orderId still missing but we have a token, check for an amending order.
        if not self._order_id and self._token:
            result = await self._page.evaluate(_FETCH_JS, [
                'GET',
                f'{_BASE}/api/order-orchestration-prod/v1/orders?size=1&sortBy=%2B&statuses=AMENDING',
                None,
                self._token,
            ])
            if result['status'] == 200:
                content = result['body'].get('content') or []
                if content:
                    self._order_id = content[0].get('customerOrderId', '')

    async def _fetch(self, method: str, url: str, body=None) -> dict:
        # Pick up any token the SPA's fetch hook has captured since last call.
        fresh = await self._page.evaluate("() => window.__shopmeToken__ || ''")
        if fresh:
            self._token = fresh

        result = await self._page.evaluate(_FETCH_JS, [method, url, body, self._token])

        if result['status'] == 401:
            # Token expired — reload to get a fresh SSR token, then retry once.
            await self._page.reload(wait_until='networkidle')
            await self._init_context()
            result = await self._page.evaluate(_FETCH_JS, [method, url, body, self._token])

        return result

    async def _gql(self, query: str, variables: dict) -> dict:
        result = await self._fetch('POST', _GRAPHQL, {'query': query, 'variables': variables})
        if result['status'] != 200:
            raise RuntimeError(f'GraphQL HTTP {result["status"]}')
        data = result['body']
        if data.get('errors'):
            raise RuntimeError(', '.join(e['message'] for e in data['errors']))
        return data['data']

    async def _lookup_names(self, line_numbers: list[str]) -> dict[str, dict]:
        encoded = '%2B'.join(line_numbers)
        result = await self._fetch('GET', f'{_BASE}/api/products-prod/v1/products/{encoded}?view=SUMMARY')
        if result['status'] != 200:
            return {}
        return {p['lineNumber']: {'name': p.get('name'), 'size': p.get('size')}
                for p in result['body'].get('products', [])}

    @staticmethod
    def _enc_product(line_number: str, product_id: str) -> str:
        return f'{line_number}:{product_id}'

    @staticmethod
    def _dec_product(product_id: str) -> tuple[str, str]:
        line_number, prod_id = product_id.split(':', 1)
        return line_number, prod_id

    @staticmethod
    def _enc_cart_item(trolley_item_id: int, uom: str) -> str:
        return f'{trolley_item_id}:{uom}'

    @staticmethod
    def _dec_cart_item(cart_item_id: str) -> tuple[int, str]:
        tid, uom = cart_item_id.split(':', 1)
        return int(tid), uom

    async def search(self, term: str, size: int = 10) -> list[Product]:
        if not self._customer_id:
            await self._init_context()
        url = (f'{_BASE}/api/content-prod/v2/cms/publish/productcontent'
               f'/search/{self._customer_id}?clientType=WEB_APP')
        params: dict = {
            'searchTerm': term, 'size': size, 'sortBy': 'MOST_POPULAR',
            'searchTags': [], 'filterTags': [], 'categoryLevel': 1,
        }
        if self._order_id:
            params['orderId'] = self._order_id
        body = {'customerSearchRequest': {'queryParams': params}}
        result = await self._fetch('POST', url, body)
        if result['status'] != 200:
            raise RuntimeError(f'Search HTTP {result["status"]}')
        products = []
        for c in result['body'].get('componentsAndProducts', []):
            p = c.get('searchProduct')
            if not p:
                continue
            promo = p.get('promotion')
            products.append(Product(
                id=self._enc_product(p['lineNumber'], p['id']),
                name=p.get('name', ''),
                size=p.get('size'),
                price=p.get('displayPrice', ''),
                price_per_unit=p.get('displayPriceQualifier'),
                promotion=promo.get('promotionDescription') if promo else None,
            ))
        return products

    async def get_cart(self) -> Cart:
        if not self._order_id:
            await self._init_context()
        data = await self._gql("""
            query($orderId: ID!) {
              getTrolley(orderId: $orderId) {
                trolley {
                  trolleyItems {
                    trolleyItemId lineNumber productId
                    quantity { amount uom }
                    totalPrice { amount currencyCode }
                  }
                  trolleyTotals {
                    itemTotalEstimatedCost { amount currencyCode }
                    savingsFromOffers { amount currencyCode }
                  }
                }
              }
            }""", {'orderId': self._order_id})

        trolley = (data.get('getTrolley') or {}).get('trolley') or {}
        raw_items = trolley.get('trolleyItems') or []

        name_map = {}
        if raw_items:
            name_map = await self._lookup_names([i['lineNumber'] for i in raw_items])

        items = []
        for i in raw_items:
            info = name_map.get(i['lineNumber'], {})
            items.append(CartItem(
                cart_item_id=self._enc_cart_item(i['trolleyItemId'], i['quantity']['uom']),
                product_id=self._enc_product(i['lineNumber'], i['productId']),
                name=info.get('name') or '',
                qty=i['quantity']['amount'],
                price=f"£{i['totalPrice']['amount']:.2f}",
            ))

        totals = trolley.get('trolleyTotals') or {}
        total_amt = (totals.get('itemTotalEstimatedCost') or {}).get('amount', 0)
        savings_amt = (totals.get('savingsFromOffers') or {}).get('amount', 0)
        return Cart(
            items=items,
            total=f'£{total_amt:.2f}',
            savings=f'£{savings_amt:.2f}' if savings_amt > 0 else None,
        )

    async def add(self, product_id: str, qty: int = 1) -> Cart:
        if not self._order_id:
            await self._init_context()
        line_number, prod_id = self._dec_product(product_id)
        data = await self._gql("""
            mutation($orderId: ID!, $trolleyItem: TrolleyItemInput) {
              addItemToTrolley(orderId: $orderId, trolleyItem: $trolleyItem) {
                trolley { trolleyTotals { itemTotalEstimatedCost { amount currencyCode } } }
                failures { message type }
              }
            }""", {
            'orderId': self._order_id,
            'trolleyItem': {
                'lineNumber': line_number,
                'productId': prod_id,
                'quantity': {'amount': qty, 'uom': 'C62'},
                'trolleyItemId': -int(line_number),
            },
        })
        failures = (data.get('addItemToTrolley') or {}).get('failures') or []
        if failures:
            raise RuntimeError(', '.join(f['message'] for f in failures))
        return await self.get_cart()

    async def set_qty(self, cart_item_id: str, qty: int) -> Cart:
        if not self._order_id:
            await self._init_context()
        trolley_item_id, uom = self._dec_cart_item(cart_item_id)

        cart = await self.get_cart()
        item = next((i for i in cart.items if i.cart_item_id == cart_item_id), None)
        if item is None:
            raise RuntimeError(f'Cart item {cart_item_id!r} not found')
        line_number, prod_id = self._dec_product(item.product_id)

        data = await self._gql("""
            mutation($orderId: ID!, $trolleyItem: TrolleyItemInput) {
              updateTrolleyItem(orderId: $orderId, trolleyItem: $trolleyItem) {
                trolley { trolleyTotals { itemTotalEstimatedCost { amount currencyCode } } }
                failures { message type }
              }
            }""", {
            'orderId': self._order_id,
            'trolleyItem': {
                'trolleyItemId': trolley_item_id,
                'lineNumber': line_number,
                'productId': prod_id,
                'quantity': {'amount': qty, 'uom': uom},
                'canSubstitute': True,
                'personalisedMessage': None,
            },
        })
        failures = (data.get('updateTrolleyItem') or {}).get('failures') or []
        if failures:
            raise RuntimeError(', '.join(f['message'] for f in failures))
        return await self.get_cart()

    async def clear(self) -> Cart:
        if not self._order_id:
            await self._init_context()
        await self._gql(
            'mutation($orderId: ID!) { emptyTrolley(orderId: $orderId) { trolley { orderId } } }',
            {'orderId': self._order_id},
        )
        ctx = await self._page.evaluate("() => localStorage.getItem('wtr_order_id')")
        if ctx:
            self._order_id = ctx
        return await self.get_cart()

    async def get_orders(self, size: int = 15) -> list[Order]:
        url = f'{_BASE}/api/order-orchestration-prod/v1/orders?size={size}&sortBy=%2B'
        result = await self._fetch('GET', url)
        if result['status'] != 200:
            raise RuntimeError(f'Orders HTTP {result["status"]}')
        orders = []
        for o in result['body'].get('content', []):
            total = ((o.get('totals') or {}).get('estimated') or {}).get('totalPrice', {}).get('amount')
            slots = o.get('slots') or []
            orders.append(Order(
                order_id=o['customerOrderId'],
                status=o['status'],
                placed_date=o.get('created'),
                delivery_date=slots[0].get('startDateTime') if slots else None,
                total=f'£{float(total):.2f}' if total is not None else None,
                item_count=o.get('numberOfItems'),
            ))
        return orders

    async def get_order(self, order_id: str) -> OrderDetail:
        result = await self._fetch('GET', f'{_BASE}/api/order-orchestration-prod/v1/orders/{order_id}')
        if result['status'] != 200:
            raise RuntimeError(f'Order detail HTTP {result["status"]}')
        order = result['body']
        raw_items = order.get('orderLines') or []

        name_map = {}
        if raw_items:
            name_map = await self._lookup_names([l['lineNumber'] for l in raw_items])

        items = []
        for l in raw_items:
            info = name_map.get(l['lineNumber'], {})
            unit = (l.get('estimatedUnitPrice') or {}).get('amount')
            total = (l.get('estimatedTotalPrice') or {}).get('amount')
            items.append(OrderItem(
                line_number=l['lineNumber'],
                name=info.get('name'),
                size=info.get('size'),
                qty=(l.get('quantity') or {}).get('amount'),
                unit_price=f'£{float(unit):.2f}' if unit is not None else None,
                total_price=f'£{float(total):.2f}' if total is not None else None,
            ))

        grand_total = ((order.get('totals') or {}).get('estimated') or {}).get('totalPrice', {}).get('amount')
        slots = order.get('slots') or []
        return OrderDetail(
            order_id=order.get('customerOrderId', order_id),
            status=order.get('status', ''),
            placed_date=order.get('created'),
            delivery_date=slots[0].get('startDateTime') if slots else None,
            total=f'£{float(grand_total):.2f}' if grand_total is not None else None,
            items=items,
        )

File: packages/grocery-core/src/vendors/waitrose.ts
import type { Page } from "playwright-core";
import type { Cart, CartItem, Order, OrderDetail, OrderItem, Product } from "../models.js";
import type { ShoppingVendor } from "./base.js";

type AnyRecord = Record<string, any>;

const BASE = "https://www.waitrose.com";
const GRAPHQL = `${BASE}/api/graphql-prod/graph/live`;
const POUND = "\u00a3";

interface FetchResult extends Record<string, unknown> {
  status: number;
  body: any;
}

export function encodeWaitroseProductId(lineNumber: string, productId: string): string {
  return `${lineNumber}:${productId}`;
}

export function decodeWaitroseProductId(productId: string): [string, string] {
  const splitAt = productId.indexOf(":");
  if (splitAt === -1) {
    return [productId, ""];
  }

  return [productId.slice(0, splitAt), productId.slice(splitAt + 1)];
}

export function encodeWaitroseCartItemId(trolleyItemId: number, uom: string): string {
  return `${trolleyItemId}:${uom}`;
}

export function decodeWaitroseCartItemId(cartItemId: string): [number, string] {
  const splitAt = cartItemId.indexOf(":");
  if (splitAt === -1) {
    return [Number.parseInt(cartItemId, 10), ""];
  }

  return [
    Number.parseInt(cartItemId.slice(0, splitAt), 10),
    cartItemId.slice(splitAt + 1)
  ];
}

export class WaitroseVendor implements ShoppingVendor {
  private customerId = "";
  private orderId = "";
  private token = "";

  constructor(private readonly page: Page) {}

  static _enc_product(lineNumber: string, productId: string): string {
    return encodeWaitroseProductId(lineNumber, productId);
  }

  static _dec_product(productId: string): [string, string] {
    return decodeWaitroseProductId(productId);
  }

  static _enc_cart_item(trolleyItemId: number, uom: string): string {
    return encodeWaitroseCartItemId(trolleyItemId, uom);
  }

  static _dec_cart_item(cartItemId: string): [number, string] {
    return decodeWaitroseCartItemId(cartItemId);
  }

  async initContext(): Promise<void> {
    const ctx = await this.page.evaluate(() => {
      const shopWindow = window as typeof window & {
        __shopmeHooked?: boolean;
        __shopmeToken__?: string;
      };

      if (!shopWindow.__shopmeHooked) {
        shopWindow.__shopmeHooked = true;
        const orig = window.fetch;
        window.fetch = function(input, init) {
          init = init || {};
          const h = (init.headers || {}) as any;
          const auth = (typeof h.get === 'function')
            ? h.get('authorization')
            : (h.authorization || h.Authorization || '');
          if (auth && auth.startsWith('Bearer ')) shopWindow.__shopmeToken__ = auth;
          return orig.call(this, input, init);
        };
      }

      const clean = (value: unknown): string => {
        const text = typeof value === 'string' ? value : '';
        return text === 'undefined' || text === 'null' ? '' : text;
      };

      const ctx = {
        customerId: clean(localStorage.getItem('wtr_customer_id')),
        orderId: clean(localStorage.getItem('wtr_order_id')),
        token: clean(shopWindow.__shopmeToken__)
      };

      for (const script of Array.from(document.querySelectorAll('script'))) {
        const text = script.textContent || '';
        if (!ctx.token) {
          const m = text.match(/"accessToken":"(Bearer [^"]+)"/);
          if (m) ctx.token = clean(m[1]);
        }
        if (!ctx.orderId) {
          const m = text.match(/"customerOrderId":"([^"]+)"/);
          if (m) ctx.orderId = clean(m[1]);
        }
        if (!ctx.customerId) {
          const m = text.match(/"customerId":"([^"]+)"/);
          if (m) ctx.customerId = clean(m[1]);
        }
        if (ctx.token && ctx.orderId && ctx.customerId) break;
      }

      return ctx;
    }) as AnyRecord;

    this.customerId = String(ctx.customerId ?? "");
    this.orderId = String(ctx.orderId ?? "");
    this.token = String(ctx.token ?? "");

    if (!this.orderId && this.token) {
      const result = await this.fetchJson(
        "GET",
        `${BASE}/api/order-orchestration-prod/v1/orders?size=1&sortBy=%2B&statuses=AMENDING`,
        null
      );
      if (result.status === 200) {
        const content = Array.isArray(result.body?.content) ? result.body.content : [];
        if (content.length > 0) {
          this.orderId = String(content[0]?.customerOrderId ?? "");
        }
      }
    }
  }

  async rawFetch(method: string, url: string, body: unknown = null): Promise<Record<string, unknown>> {
    return this.fetchJson(method, url, body);
  }

  async search(term: string, size = 10): Promise<Product[]> {
    if (!this.customerId) {
      await this.initContext();
    }

    const url = `${BASE}/api/content-prod/v2/cms/publish/productcontent/search/${this.customerId}?clientType=WEB_APP`;
    const params: AnyRecord = {
      searchTerm: term,
      size,
      sortBy: "MOST_POPULAR",
      searchTags: [],
      filterTags: [],
      categoryLevel: 1
    };
    if (this.orderId) {
      params.orderId = this.orderId;
    }

    const result = await this.fetchJson("POST", url, {
      customerSearchRequest: { queryParams: params }
    });
    if (result.status !== 200) {
      throw new Error(`Search HTTP ${result.status}`);
    }

    const products: Product[] = [];
    for (const component of arrayOfRecords(result.body?.componentsAndProducts)) {
      const product = component.searchProduct;
      if (!product) {
        continue;
      }
      const promo = product.promotion;
      products.push({
        id: encodeWaitroseProductId(String(product.lineNumber), String(product.id)),
        name: String(product.name ?? ""),
        size: stringOrNull(product.size),
        price: String(product.displayPrice ?? ""),
        price_per_unit: stringOrNull(product.displayPriceQualifier),
        promotion: promo ? stringOrNull(promo.promotionDescription) : null
      });
    }

    return products;
  }

  async getCart(): Promise<Cart> {
    if (!this.orderId) {
      await this.initContext();
    }

    const data = await this.gql(`
      query($orderId: ID!) {
        getTrolley(orderId: $orderId) {
          trolley {
            trolleyItems {
              trolleyItemId lineNumber productId
              quantity { amount uom }
              totalPrice { amount currencyCode }
            }
            trolleyTotals {
              itemTotalEstimatedCost { amount currencyCode }
              savingsFromOffers { amount currencyCode }
            }
          }
        }
      }`, { orderId: this.orderId });

    const trolley = data?.getTrolley?.trolley ?? {};
    const rawItems = arrayOfRecords(trolley.trolleyItems);
    const nameMap = rawItems.length > 0
      ? await this.lookupNames(rawItems.map((item) => String(item.lineNumber)))
      : {};

    const items: CartItem[] = rawItems.map((item) => {
      const info = nameMap[String(item.lineNumber)] ?? {};
      return {
        cart_item_id: encodeWaitroseCartItemId(Number(item.trolleyItemId), String(item.quantity?.uom ?? "")),
        product_id: encodeWaitroseProductId(String(item.lineNumber), String(item.productId)),
        name: String(info.name ?? ""),
        qty: Number(item.quantity?.amount ?? 0),
        price: money(item.totalPrice?.amount)
      };
    });

    const totals = trolley.trolleyTotals ?? {};
    const totalAmount = Number(totals.itemTotalEstimatedCost?.amount ?? 0);
    const savingsAmount = Number(totals.savingsFromOffers?.amount ?? 0);

    return {
      items,
      total: money(totalAmount),
      savings: savingsAmount > 0 ? money(savingsAmount) : null
    };
  }

  async add(productId: string, qty = 1): Promise<Cart> {
    if (!this.orderId) {
      await this.initContext();
    }

    const [lineNumber, prodId] = decodeWaitroseProductId(productId);
    const data = await this.gql(`
      mutation($orderId: ID!, $trolleyItem: TrolleyItemInput) {
        addItemToTrolley(orderId: $orderId, trolleyItem: $trolleyItem) {
          trolley { trolleyTotals { itemTotalEstimatedCost { amount currencyCode } } }
          failures { message type }
        }
      }`, {
      orderId: this.orderId,
      trolleyItem: {
        lineNumber,
        productId: prodId,
        quantity: { amount: qty, uom: "C62" },
        trolleyItemId: -Number.parseInt(lineNumber, 10)
      }
    });

    const failures = arrayOfRecords(data?.addItemToTrolley?.failures);
    if (failures.length > 0) {
      throw new Error(failures.map((failure) => failure.message).join(", "));
    }

    return this.getCart();
  }

  async setQty(cartItemId: string, qty: number): Promise<Cart> {
    if (!this.orderId) {
      await this.initContext();
    }

    const [trolleyItemId, uom] = decodeWaitroseCartItemId(cartItemId);
    const cart = await this.getCart();
    const item = cart.items.find((candidate) => candidate.cart_item_id === cartItemId);
    if (!item) {
      throw new Error(`Cart item ${JSON.stringify(cartItemId)} not found`);
    }

    const [lineNumber, prodId] = decodeWaitroseProductId(item.product_id);
    const data = await this.gql(`
      mutation($orderId: ID!, $trolleyItem: TrolleyItemInput) {
        updateTrolleyItem(orderId: $orderId, trolleyItem: $trolleyItem) {
          trolley { trolleyTotals { itemTotalEstimatedCost { amount currencyCode } } }
          failures { message type }
        }
      }`, {
      orderId: this.orderId,
      trolleyItem: {
        trolleyItemId,
        lineNumber,
        productId: prodId,
        quantity: { amount: qty, uom },
        canSubstitute: true,
        personalisedMessage: null
      }
    });

    const failures = arrayOfRecords(data?.updateTrolleyItem?.failures);
    if (failures.length > 0) {
      throw new Error(failures.map((failure) => failure.message).join(", "));
    }

    return this.getCart();
  }

  async clear(): Promise<Cart> {
    if (!this.orderId) {
      await this.initContext();
    }

    await this.gql(
      "mutation($orderId: ID!) { emptyTrolley(orderId: $orderId) { trolley { orderId } } }",
      { orderId: this.orderId }
    );
    const ctx = await this.page.evaluate(() => localStorage.getItem("wtr_order_id")) as string | null;
    if (ctx) {
      this.orderId = ctx;
    }

    return this.getCart();
  }

  async getOrders(size = 15): Promise<Order[]> {
    const result = await this.fetchJson("GET", `${BASE}/api/order-orchestration-prod/v1/orders?size=${size}&sortBy=%2B`);
    if (result.status !== 200) {
      throw new Error(`Orders HTTP ${result.status}`);
    }

    return arrayOfRecords(result.body?.content).map((order): Order => {
      const total = order.totals?.estimated?.totalPrice?.amount;
      const slots = arrayOfRecords(order.slots);
      return {
        order_id: String(order.customerOrderId),
        status: String(order.status ?? ""),
        placed_date: stringOrNull(order.created),
        delivery_date: slots.length > 0 ? stringOrNull(slots[0].startDateTime) : null,
        total: total !== undefined && total !== null ? money(Number(total)) : null,
        item_count: numberOrNull(order.numberOfItems)
      };
    });
  }

  async getOrder(orderId: string): Promise<OrderDetail> {
    const result = await this.fetchJson("GET", `${BASE}/api/order-orchestration-prod/v1/orders/${orderId}`);
    if (result.status !== 200) {
      throw new Error(`Order detail HTTP ${result.status}`);
    }

    const order = result.body ?? {};
    const rawItems = arrayOfRecords(order.orderLines);
    const nameMap = rawItems.length > 0
      ? await this.lookupNames(rawItems.map((line) => idOrNull(line.lineNumber)).filter(isString))
      : {};

    const items: OrderItem[] = rawItems.map((line) => {
      const lineNumber = idOrNull(line.lineNumber);
      const rawProductId = idOrNull(line.productId);
      const info = lineNumber ? nameMap[lineNumber] ?? {} : {};
      const unit = line.estimatedUnitPrice?.amount;
      const total = line.estimatedTotalPrice?.amount;
      return {
        product_id: lineNumber && rawProductId ? encodeWaitroseProductId(lineNumber, rawProductId) : lineNumber,
        name: stringOrNull(info.name),
        size: stringOrNull(info.size),
        qty: numberOrNull(line.quantity?.amount),
        unit_price: unit !== undefined && unit !== null ? money(Number(unit)) : null,
        total_price: total !== undefined && total !== null ? money(Number(total)) : null
      };
    });

    const grandTotal = order.totals?.estimated?.totalPrice?.amount;
    const slots = arrayOfRecords(order.slots);

    return {
      order_id: String(order.customerOrderId ?? orderId),
      status: String(order.status ?? ""),
      placed_date: stringOrNull(order.created),
      delivery_date: slots.length > 0 ? stringOrNull(slots[0].startDateTime) : null,
      total: grandTotal !== undefined && grandTotal !== null ? money(Number(grandTotal)) : null,
      items
    };
  }

  private async fetchJson(method: string, url: string, body: unknown = null): Promise<FetchResult> {
    const fresh = await this.page.evaluate(() => {
      const shopWindow = window as typeof window & { __shopmeToken__?: string };
      return shopWindow.__shopmeToken__ || "";
    }) as string;
    if (fresh) {
      this.token = fresh;
    }

    const request: [string, string, unknown, string] = [method, url, body ?? null, this.token];
    let result = await this.page.evaluate(fetchInPage, request) as FetchResult;
    if (result.status === 401) {
      await this.page.reload({ waitUntil: "domcontentloaded" });
      await this.initContext();
      const retryRequest: [string, string, unknown, string] = [method, url, body ?? null, this.token];
      result = await this.page.evaluate(fetchInPage, retryRequest) as FetchResult;
    }

    return result;
  }

  private async gql(query: string, variables: Record<string, unknown>): Promise<any> {
    const result = await this.fetchJson("POST", GRAPHQL, { query, variables });
    if (result.status !== 200) {
      throw new Error(`GraphQL HTTP ${result.status}`);
    }

    if (Array.isArray(result.body?.errors) && result.body.errors.length > 0) {
      throw new Error(result.body.errors.map((error: AnyRecord) => error.message).join(", "));
    }

    return result.body?.data;
  }

  private async lookupNames(lineNumbers: string[]): Promise<Record<string, { name?: string; size?: string }>> {
    const encoded = lineNumbers.join("%2B");
    const result = await this.fetchJson("GET", `${BASE}/api/products-prod/v1/products/${encoded}?view=SUMMARY`);
    if (result.status !== 200) {
      return {};
    }

    return Object.fromEntries(
      arrayOfRecords(result.body?.products).map((product) => [
        String(product.lineNumber),
        {
          name: stringOrNull(product.name) ?? undefined,
          size: stringOrNull(product.size) ?? undefined
        }
      ])
    );
  }
}

function arrayOfRecords(value: unknown): AnyRecord[] {
  return Array.isArray(value) ? value.filter((entry) => entry && typeof entry === "object") as AnyRecord[] : [];
}

function stringOrNull(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function idOrNull(value: unknown): string | null {
  return value === undefined || value === null || value === "" ? null : String(value);
}

function isString(value: string | null): value is string {
  return value !== null;
}

function numberOrNull(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

function money(value: unknown): string {
  return `${POUND}${Number(value ?? 0).toFixed(2)}`;
}

async function fetchInPage([method, url, body, token]: [string, string, unknown, string]): Promise<FetchResult> {
  const opts: RequestInit = {
    method,
    headers: { "Content-Type": "application/json" }
  };
  if (token) {
    (opts.headers as Record<string, string>).Authorization = token;
  }
  if (body !== null) {
    opts.body = JSON.stringify(body);
  }

  const response = await fetch(url, opts);
  const text = await response.text();
  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch {
    data = text;
  }

  return { status: response.status, body: data };
}
